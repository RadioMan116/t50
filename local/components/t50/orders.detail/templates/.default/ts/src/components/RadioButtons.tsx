import { h, Component } from "preact";

type DataItem = {value: string, title: string}
interface Props {
    data: DataItem[],
    name: string,
    value: string,
    onChange?: (value: string) => void,
    itemClass: string
}

export default class RadioButtons extends Component<Props>
{
    state: {value?: string} = {value: null}
    static counter = 0

    constructor(props:Props){
        super(props)
        this.setState({value: this.props.value})
    }

    change(dataItem: DataItem){
        this.setState({value: dataItem.value});
        if( this.props.onChange != null )
            this.props.onChange(dataItem.value);
    }

    componentWillReceiveProps(props: Props){
        this.setState({value: props.value})
    }

    getItem(dataItem: DataItem) {
        let id = [this.props.name, (++RadioButtons.counter)].join("_")
        let checked = dataItem.value == this.state.value;

        return <div class={this.props.itemClass}>
            <div class="check-elem">
                <input type="radio" id={id} class="check-elem__input" checked={checked} onClick={this.change.bind(this, dataItem)} />
                <label for={id} class="check-elem__label">{dataItem.title}</label>
            </div>
        </div>
    }

    render() {
        return <div>
            {this.props.data.map(item => {
                return this.getItem(item)
            })}
        </div>
    }
}