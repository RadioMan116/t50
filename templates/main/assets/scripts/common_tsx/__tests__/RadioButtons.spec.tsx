import { mount, shallow } from 'enzyme';
import { h } from 'preact';
import RadioButtons from "../src/RadioButtons"
import { AssertionError } from 'assert';
import { shallowToJson } from 'enzyme-to-json';

describe('RadioButtons', () => {
    it('check default has not checked', () => {
        let data = [
            {value: "Y", title: "Да"},
            {value: "N", title: "Нет"},
        ]
        const wrapper = shallow(<RadioButtons onChange={null} data={data} value="" itemClass="table__item" name="test"/>)
        let radio = wrapper.find("input");
        expect(radio.at(0).props().checked).toBe(false);
        expect(radio.at(1).props().checked).toBe(false);
    });

    it('has checked', () => {
        let data = [
            {value: "Y", title: "Да"},
            {value: "N", title: "Нет"},
        ]
        const wrapper = shallow(<RadioButtons onChange={null} data={data} value="N" itemClass="table__item" name="test"/>)
        expect(wrapper.state().value).toBe("N");
        let radio = wrapper.find("input");
        expect(radio.at(0).props().checked).toBe(false);
        expect(radio.at(1).props().checked).toBe(true);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('test callback', () => {
        const onChange = (value: string) => {
            expect(value).toBe("N");
        };
        let data = [
            {value: "Y", title: "Да"},
            {value: "N", title: "Нет"},
        ];
        const wrapper = shallow(<RadioButtons onChange={onChange} data={data} value="Y" itemClass="table__item" name="test"/>)

        expect(wrapper.state().value).toBe("Y");
        wrapper.setState({value: "N"})
        wrapper.setState({value: "Y"})
        let radio = wrapper.find("input");
        expect(radio.at(0).props().checked).toBe(true);
        expect(radio.at(1).props().checked).toBe(false);

        radio.at(1).simulate("click")
        expect(wrapper.state().value).toBe("N");

        wrapper.setState({value: "Y"})
        wrapper.setState({value: "N"})
        radio = wrapper.find("input");
        expect(radio.at(0).props().checked).toBe(false);
        expect(radio.at(1).props().checked).toBe(true);
    });

    it('change props', () => {
        let data = [
            {value: "Y", title: "Да"},
            {value: "N", title: "Нет"},
        ]
        const wrapper = shallow(<RadioButtons onChange={null} data={data} value="N" itemClass="table__item" name="test"/>)
        // console.log(wrapper.debug());
        expect(wrapper.state().value).toBe("N");
        let radio = wrapper.find("input");

        expect(radio.at(0).prop("checked")).toBe(false);
        expect(radio.at(1).prop("checked")).toBe(true);

        wrapper.setProps({ value: 'Y' });
        expect(wrapper.state().value).toBe("Y");
        // console.log(wrapper.debug());
        radio = wrapper.find("input");
        expect(radio.at(0).prop("checked")).toBe(true);
        expect(radio.at(1).prop("checked")).toBe(false);
    });

});