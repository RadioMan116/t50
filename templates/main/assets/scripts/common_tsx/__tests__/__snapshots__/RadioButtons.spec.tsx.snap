// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtons has checked 1`] = `
<div>
  <div
    className="table__item"
  >
    <div
      className="check-elem"
    >
      <input
        checked={false}
        className="check-elem__input"
        id="test_3"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="check-elem__label"
        for="test_3"
      >
        Да
      </label>
    </div>
  </div>
  <div
    className="table__item"
  >
    <div
      className="check-elem"
    >
      <input
        checked={true}
        className="check-elem__input"
        id="test_4"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="check-elem__label"
        for="test_4"
      >
        Нет
      </label>
    </div>
  </div>
</div>
`;
